torch/profiler/_utils.py

```diff
@@ -194,6 +203,86 @@ def compute_idle_time(self):
             self.metrics[EventKey(event)].idle_time_ns = EventKey(
                 event).intervals_overlap(idle_intervals)
 
+    def rank_events(self, length):
+        '''
+        Filter and Rank the events based on some heuristics:
+        1) Events that are in the falling phase of the queue depth.
+        2) Events that have a high idle_time, self_time difference.
+
+        Parameters:
+            length: The number of events to return.
+        '''
+
+        # Find the interval when qd is falling to 0
+        queue_depth_list = list(reversed(self.queue_depth_list))
+        qd_values = [e.queue_depth for e in queue_depth_list]
+
+        decrease_interval = []
+        for i in range(len(qd_values) - 1):
+            if qd_values[i] <= 1:
+                # Find next zero and if the max value between them exceeds
+                # the threshold, then we have a falling interval
+                for j in range(i + 1, len(qd_values)):
+                    if qd_values[j] <= 1:
+                        peak_idx = argmax(qd_values, start=i + 1, end=j)
+                        if peak_idx is None:
+                            continue
+                        # check for threshold
+                        if qd_values[peak_idx] - qd_values[i] > 3:
+                            decrease_interval.append(
+                                Interval(queue_depth_list[peak_idx].start,
+                                         queue_depth_list[i].start))
+                            i = j
+                            break
+
+        # Filter out events that are not in the decrease interval
+        event_list = [
+            event for event in self.metrics.keys()
+            if event.intervals_overlap(decrease_interval)
+        ]
+        if event_list:
+            self_time = torch.tensor(
+                [self.metrics[event].self_time_ns for event in event_list],
+                dtype=torch.float32)
+            idle_time = torch.tensor(
+                [self.metrics[event].idle_time_ns for event in event_list],
+                dtype=torch.float32)
+
+            normalized_gain = (idle_time -
+                               torch.mean(idle_time)) / torch.std(idle_time)
+            normalized_self = (self_time -
+                               torch.mean(self_time)) / torch.std(self_time)
+            heuristic_score_list = normalized_gain + normalized_self
+
+            # Sort events by heuristic
+            event_list = [
+                event
+                for _, event in sorted(zip(heuristic_score_list, event_list),
+                                       key=lambda x: x[0],
+                                       reverse=True)
+            ]
+            event_list = event_list[:length]
+        return event_list
+
+    def get_optimizable_events(self,
+                               length: int = 1,
+                               print_enable: bool = True):
+        event_list = self.rank_events(length)
```

Comment:
We're no longer using a user provided function?

Reply:
I guess since the way that the score is calculated is so involved (using lambda is probably unlikely), the best way for user to define their own function is to inherit from BasicEvaluation, and overwrite the rank_events() ?

Reply:
Or have an abstract base class that defines `__call__`. (Which also lets functors easily define their own helper methods and maintain state.)

Reply:
I don't quite understand the details here. Is it the case that we let user inherit the ABC, and the __call__ replaces the current rank_events() ? so just a name change?